operator v1: does not return quiescent if replicas not synced

birdayz opened the pull request:
we had cases, where quiescent condition is given, even if replicas were not synced. this caused our "wait" job to not wait anymore, even if replicas are not yet up.

See https://redpandadata.slack.com/archives/C077JRCJ807/p1725902281802229

RafalKorepta reviewed (approved):
Usage of quiescent condition to indicate what is already in the status seems wrong to me. Why the waiter can not compare `CurrentReplicas` and `ReadyReplicas`?

Never the less to not create more burden on cloud I'm fine with that change.

It would be good to add tests.

birdayz commented:
hm, i understand that point of view.
however, this shifts the burden on the caller. why does the caller need to know which fields to compare? especially, this list of fields to compare may change over time.
the reason why we add this condition is, there's a single field that can tell me, that the manifest is synced.

RafalKorepta commented:
I'm not blocking this change. It's my personal view point that may not reflect wider audiance. I'm fine merging this as is. Maybe additional tests could good to have

birdayz commented:
I will add a test. Thank you.

birdayz commented:
I could also see the value of a separate Condition, ReplicasSynced.

- OperatorQuiescent indicates that the operator itself has no outstanding work to do. this can still mean, that replicas are not yet all up - this is managed by statefulset controller, not operator directly (operator has nothing to do here)
- ReplicasSynced would indicate, all the desired replicas are up and running.

birdayz commented:
@RafalKorepta i did another pass, and added a test.
also added another change, that is slightly related: https://github.com/redpanda-data/redpanda-operator/pull/222/commits/d6816e8cf4d8dafd22ab0b0fc85cac5e027f5e36

please have another look. thank you.

chrisseto commented on src/go/k8s/internal/controller/redpanda/cluster_controller.go:
```diff
@@ -1158,6 +1158,13 @@ func getQuiescentCondition(redpandaCluster *vectorizedv1alpha1.Cluster) vectoriz
 		return condition
 	}
 
+	if redpandaCluster.Status.CurrentReplicas != redpandaCluster.Status.Replicas || redpandaCluster.Status.Replicas != redpandaCluster.Status.ReadyReplicas || (redpandaCluster.Spec.Replicas != nil && *redpandaCluster.Spec.Replicas != redpandaCluster.Status.Replicas) {
```
This is a bit of a beast to read and reason about. Could we break this into a couple variables first?
```go
hasCurrentReplicas := redpandaCluster.Status.CurrentReplicas == redpandaCluster.Status.Replicas
// ...
if !(hasCurrentReplicas && hasReadyReplicas /* ... */) {

}
```

birdayz replied:
refactored

chrisseto commented on src/go/k8s/internal/controller/redpanda/cluster_controller.go:
```diff
@@ -1158,6 +1158,13 @@ func getQuiescentCondition(redpandaCluster *vectorizedv1alpha1.Cluster) vectoriz
+	if redpandaCluster.Status.CurrentReplicas != redpandaCluster.Status.Replicas || redpandaCluster.Status.Replicas != redpandaCluster.Status.ReadyReplicas || (redpandaCluster.Spec.Replicas != nil && *redpandaCluster.Spec.Replicas != redpandaCluster.Status.Replicas) {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = "ReplicasNotSynced"
+		condition.Message = fmt.Sprintf("Replicas are not synced. spec.replicas=%d status.currentReplicas=%d,status.replicas=%d,status.readyReplicas=%d. All must be equal.", ptr.Deref(redpandaCluster.Spec.Replicas, 0), redpandaCluster.Status.CurrentReplicas, redpandaCluster.Status.Replicas, redpandaCluster.Status.ReadyReplicas)
```
If Replicas is wrapped in a `ptr.Deref` here shouldn't it be wrapped in the above condition as well?

birdayz replied:
in above condition i do a nil check

birdayz replied:
i now changed both to use ptr.Deref to treat nil as 0 here.
with NodePools, this logic has to change a bit, because nil will be a quite common and legitimate value. but it's not a problem at the moment
